File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.js

function solution(triangle) {
    let maxDepth = triangle.length
    let dp = triangle.map(layer => layer.map(element => 0))
    dp[0][0] = triangle[0][0]
    for(let currentDepth = 1 ; currentDepth < maxDepth ; currentDepth ++){
        triangle[currentDepth].forEach((element,index) => {
            let currentVal = triangle[currentDepth][index]
            if(index === 0 ) {
                dp[currentDepth][index] = currentVal + dp[currentDepth-1][index] 
            }
            else if(index === currentDepth){
                dp[currentDepth][index] = currentVal + dp[currentDepth-1][index-1]
            }
            else{
                let parentVal = getMax(dp[currentDepth-1][index-1], dp[currentDepth-1][index])
                dp[currentDepth][index] = currentVal + parentVal
            }
        })
    }
    let answer = dp[maxDepth-1].reduce((max,current) => max <= current ? current : max , 0)
    return answer
}
function getMax (a,b){
    return a >= b ? a : b
}
 // function dfs(currentDepth,currentIndex, accumVal){
//     let currentVal = triangle[currentDepth][currentIndex]
//     if(currentDepth == maxDepth-1){
//         answer = Math.max(answer,accumVal + currentVal)
//     }
//     else{
//         dfs(currentDepth+1,currentIndex,accumVal + currentVal)
//         dfs(currentDepth+1,currentIndex+1,accumVal + currentVal)
//     }
// }

File: 프로그래머스/3/43105. 정수 삼각형/정수 삼각형.java
import java.util.*;
class Solution {
    public int solution(int[][] triangle) {
        int answer = 0;
        int depth = triangle.length;
        int[][] maximum = new int[depth][];
        for(int i = 0 ; i<depth ; i++){
            maximum[i] = new int[i+1];
        }
        maximum[0][0] = triangle[0][0];
        for(int i = 0 ; i<depth-1 ; i++){
            for(int j= 0 ; j<=i ; j++){
                maximum[i+1][j] = Math.max(maximum[i][j] + triangle[i+1][j] , maximum[i+1][j]);
                maximum[i+1][j+1] = Math.max(maximum[i][j] + triangle[i+1][j+1], maximum[i+1][j+1]);
                
            }
        }
        
        return Arrays.stream(maximum[depth-1]).max().getAsInt();
    }
}
